File: fsl-blog-scala/src/main/scala/oxconcurreny/ProducerConsumer.scala
package oxconcurreny

import org.apache.commons.lang3.RandomStringUtils
import ox._
import ox.channels._

import scala.annotation.tailrec
import scala.util.Random

object ProducerConsumer {

  def getRandomTime: Int = Random.between(500, 1001)

  def producer(id: Long, queue: Sink[String])(using Ox): Fork[Unit] = fork {
    produce(id, queue)
  }

  @tailrec
  def produce(id: Long, queue: Sink[String]): Unit =
    Thread.sleep(getRandomTime)

    val value = RandomStringUtils.randomAlphanumeric(10)
    queue.send(value)
    println("producer " + id + " producing " + value)

    produce(id, queue)

  def consumer(id: Long, queue: Source[String])(using Ox): Fork[Unit] = fork {
    consume(id, queue)
  }

  @tailrec
  def consume(id: Long, queue: Source[String]): Unit =
    Thread.sleep(getRandomTime)

    val value = queue.receive()
    println("consumer " + id + " consuming value " + value)

    consume(id, queue)

  @main def main(): Unit =
    supervised {
      val queue = Channel.unlimited[String]

      val producers = List.range(1, 11).map(producer(_, queue))
      val consumers = List.range(1, 11).map(consumer(_, queue))

      (producers ++ consumers).foreach(_.join())
    }

}

File: fsl-blog-kotlin/src/main/kotlin/coroutines/ProducerConsumer.kt
package coroutines

import kotlinx.coroutines.*
import kotlinx.coroutines.channels.*
import kotlinx.coroutines.channels.Channel.Factory.UNLIMITED
import org.apache.commons.lang3.RandomStringUtils

fun getRandomTime(): Long = (500..1000).random().toLong()

suspend fun producer(id: Int, queue: SendChannel<String>) {
  while (true) {
    delay(getRandomTime())

    val value = RandomStringUtils.randomAlphanumeric(10)
    println("producer $id producing $value")
    queue.send(value)
  }
}

suspend fun consumer(id: Int, queue: ReceiveChannel<String>) {
  while (true) {
    delay(getRandomTime())

    val value = queue.receive()
    println("consumer $id consuming value $value")
  }
}

suspend fun main(args: Array<String>) {
  coroutineScope {
    val queue = Channel<String>(UNLIMITED)

    repeat(10) { i ->
      launch { producer(i + 1, queue) }
    }

    repeat(10) { i ->
      launch { consumer(i + 1, queue) }
    }
  }
}
